### PR TITLE
Migration to Nom5

### LINKED ISSUE #83: Move to nom 5.0
We're currently on `4.2.3`, whereas upstream has been bumped to 5.0. While I think the changes are mainly cosmetic/mechanical between the two, it still requires some work.

### PR DESCRIPTION
**Pull request by AnderEnder**
Migration was done in 2 parts:
- migration to nom 5 procedure macros
- migration to nom 5 functions

Some missing macros were replaced by the combination of another macros:
- **take_until_and_consume** ->  **take_until** + **take**
- **alt_complete** -> **complete** + **alt**

Missing variant **ErrorKind::Custom** was replaced with **ErrorKind::Alpha**

Some macros were replaced by the combination of functions:
- **do_parse** -> **map** + **tuple**
- **parse_to** -> **map_res** + **u64::from_str** + **map_res** + **str::from_utf8**

Closes #83

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -34,7 +34,7 @@ default = [ "serialization", "sync" ]
 num-traits = "0.2"
 byteorder = "1.0.0"
 flate2 = { version = "1.0", optional = true }
-nom = { version = "^4.0", optional = true }
+nom = { version = "^5.0", optional = true }
```
Comment (jonhoo) [inline Cargo.toml]: This can just be `5.0`, no need for the `^`.

File: src/serialization/interval_log/mod.rs [hunk 2]
```diff
@@ -674,61 +683,83 @@ fn system_time_as_fp_seconds(time: time::SystemTime) -> f64 {
...
-    do_parse!(tag!("#") >> take_until_and_consume!("\n") >> (()))
-);
-
-named!(legend<&[u8], ()>,
-    do_parse!(tag!("\"StartTimestamp\"") >> take_until_and_consume!("\n") >> (()))
-);
-
-named!(ignored_line<&[u8], ()>, alt!(comment_line | legend));
+fn start_time(input: &[u8]) -> IResult<&[u8], LogEntry, (&[u8], nom::error::ErrorKind)> {
+    map(
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: I _think_ the new idiomatic way to do these in `nom 5.0` is not to use combinators, but to instead write straight code like this:
https://github.com/jonhoo/msql-srv/blob/5fd1c7573842a94a2bc5be170c771a30118cc250/src/commands.rs#L13-L21

File: src/serialization/interval_log/mod.rs [hunk 3]
```diff
@@ -674,61 +683,83 @@ fn system_time_as_fp_seconds(time: time::SystemTime) -> f64 {
...
+        tuple((
+            opt(map(
+                map_res(
+                    map(tuple((tag("Tag="), take_until(","), take(1_usize))), |p| {
+                        p.1
+                    }),
+                    str::from_utf8,
+                ),
+                |s| Tag(s),
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: I think this can actually just be `Tag`

File: src/serialization/interval_log/mod.rs [hunk 4]
```diff
@@ -674,61 +683,83 @@ fn system_time_as_fp_seconds(time: time::SystemTime) -> f64 {
...
+        |x| LogEntry::BaseTime(x.1),
+    )(input)
+}
+
+fn interval_hist(input: &[u8]) -> IResult<&[u8], LogEntry, (&[u8], nom::error::ErrorKind)> {
+    map(
+        tuple((
+            opt(map(
+                map_res(
+                    map(tuple((tag("Tag="), take_until(","), take(1_usize))), |p| {
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: Not sure I follow where the `take(1_usize)` comes from?

File: src/serialization/interval_log/mod.rs [hunk 5]
```diff
@@ -674,61 +683,83 @@ fn system_time_as_fp_seconds(time: time::SystemTime) -> f64 {
...
+                |s| Tag(s),
+            )),
+            fract_sec_duration,
+            char(','),
+            fract_sec_duration,
+            char(','),
+            double,
+            char(','),
+            map_res(take_until("\n"), str::from_utf8),
+            take(1_usize),
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: Same with this ­— why `1_usize` specifically?

File: src/serialization/interval_log/mod.rs (hunk 4 above)
Comment (jonhoo) [reply src/serialization/interval_log/mod.rs]: Ohh, it's to also consume the newline!

File: src/serialization/interval_log/mod.rs [hunk 6]
```diff
@@ -753,37 +784,26 @@ fn fract_sec_duration(input: &[u8]) -> IResult<&[u8], time::Duration> {
...
-    )
-);
+fn fract_sec_tuple(input: &[u8]) -> IResult<&[u8], (u64, &str), (&[u8], nom::error::ErrorKind)> {
+    map(
+        tuple((
+            map_res(
+                map_res(recognize(take_until(".")), str::from_utf8),
+                u64::from_str,
+            ),
+            recognize(take_until(".")),
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: Not sure I understand why this is here?

**Comment by jonhoo** [comment]
Ah, great work, thanks! I left some inline comments.

File: src/serialization/interval_log/mod.rs (hunk 5 above)
Comment (AnderEnder) [reply src/serialization/interval_log/mod.rs]:
```
error[E0277]: the trait bound `i32: nom::ToUsize` is not satisfied
   --> src/serialization/interval_log/mod.rs:731:33
    |
731 | ...                   take(1)
    |                       ^^^^ the trait `nom::ToUsize` is not implemented for `i32`
```

File: src/serialization/interval_log/mod.rs (hunk 6 above)
Comment (AnderEnder) [reply src/serialization/interval_log/mod.rs]: I am not as well :)

**Comment by marshallpierce** [comment]
Thanks for cleaning up my old nom macro mess; this is a lot easier to read. :)

File: src/serialization/interval_log/mod.rs [hunk 7]
```diff
@@ -683,81 +683,75 @@ fn system_time_as_fp_seconds(time: time::SystemTime) -> f64 {
...
-                start_timestamp,
-                duration,
-                max,
-                encoded_histogram,
-            })
-        },
-    )(input)
+fn interval_hist(input: &[u8]) -> IResult<&[u8], LogEntry, (&[u8], ErrorKind)> {
+    let (input, tag) = opt(map(
+        map_res(
```
Comment (jonhoo) [inline src/serialization/interval_log/mod.rs]: I'd probably personally write a helper function for the inner parser here, but :man_shrugging:

File: src/serialization/interval_log/mod.rs (hunk 6 above)
Comment (jonhoo) [reply src/serialization/interval_log/mod.rs]: @marshallpierce Do you remember? :p

Comment (marshallpierce) [reply src/serialization/interval_log/mod.rs]: Oof, wish I did. There was some quirk of formatting that it was dealing with, like `4.` with the trailing dot or something like that.

Comment (marshallpierce) [reply src/serialization/interval_log/mod.rs]: I think at the time I was really struggling with nom macros so _everything_ seemed confusing and I didn't think to comment that part :'(

**Comment by jonhoo** [comment]
Looks like CI is failing?

**Comment by AnderEnder** [comment]
@jonhoo 
Looks like **tarpaulin** released new feature to support coverage of examples:
https://github.com/xd009642/tarpaulin/compare/0.9.1...0.9.2

I am not yet sure if it is a **tarpaulin** issue or some additional configuration is required.

**Pull request merged**